File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        ArrayDeque st = new ArrayDeque();
        HashMap<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');

        for (var x : s.toCharArray()) {
            if (map.containsKey(x)) {
                if (st.peek() == map.get(x)) {
                    st.pop();
                }
                else {
                    return false;
                }   
            }
            else {
                st.push(x);
            }
        }

        return st.size() == 0;
    }
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        map = {')' : '(',
        ']':'[',
        '}':'{'}

        for i in range(len(s)):
            if s[i] in map:
                if stack and stack[-1] != map[s[i]]:
                    return False
                if stack:
                    stack.pop()
                else:
                    return False
                continue
            stack.append(s[i])
            # print(stack)
        return len(stack) == 0